### PR TITLE
CSD-12 Opportunities filters built with React and PDB

### PR DESCRIPTION
**Pull request by pookmish**
# READY FOR REVIEW

# Summary
- Search API view for opportunities results
- Progressively decoupled blocks module to build a custom filter UI for the user
- add some more fields to the opportunity content type with taxonomy vocabs

# Need Review By (Date)
- 4/29

# Urgency
- medium

# Steps to Test
1. I set up a page on tugboat for quick verification: https://pr8-duzhkv3nisyijofdbmuqrmfsemrk7qjz.tugboat.qa/search-opportunities
1. `composer require SU-HKKU/cardinal_service_profile:dev-CSD-12-filters-react`
1. `blt drupal:install --site=cardinalservice` or `blt drupal:update --site=cardinalservice` if you already have a site.
1. create some opportunity content via devel, manually, or `drush @cardinalservice.local mim solo_opportunities`
1. create an empty page
1. edit the layout and add the "Opportunities Filtering List" block & the "Opportunities: All Filtered" blocks
1. after saving the layout, verify the filters function and you see some results (styling to come later)

# See Also
- [PR Checklist](https://gist.github.com/sherakama/0ba17601381e3adbe0cad566ad4d80a5)

### CODE REVIEW

**Comment by jbickar** [comment]
Started on this today. Might get to it this afternoon.

**Comment by jbickar** [comment]
I don't have a "Opportunities Search API" block available. Something not enabled?

**Comment by pookmish** [comment]
woops. forgot to change the test steps "Opportunities: All Filtered" is the block

**Comment by sherakama** [comment]
Will review this later today.

**Comment by pookmish** [comment]
this is starting to be a blocker for further work that relies on it

**Comment by jbickar** [comment]
OK, thanks for the ping.

**Comment by sherakama** [comment]
I forgot to review it yesterday. I'll make it a priority today.

File: modules/cardinal_service_rest/cardinal_service_rest.info.yml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+name: Cardinal Service Rest
+description: Rest API endpoints for Cardinal Service.
+core_version_requirement: ^8.8 || ^9
+type: module
```
Comment (sherakama) [inline modules/cardinal_service_rest/cardinal_service_rest.info.yml:4]: Missing version and dependencies keys.

File: themes/cardinal_service/cardinal_service.info.yml [hunk 2]
```diff
@@ -30,3 +30,10 @@ libraries:
 
 ckeditor_stylesheets:
   - dist/css/ckeditor.css
+
+libraries-override:
+  # Replace an asset with another.
+  pdb_react/react:
+    js:
+      //cdnjs.cloudflare.com/ajax/libs/react/15.0.1/react.min.js: //unpkg.com/react@16/umd/react.production.min.js
+      //cdnjs.cloudflare.com/ajax/libs/react/15.0.1/react-dom.min.js: //unpkg.com/react-dom@16/umd/react-dom.production.min.js
```
Comment (sherakama) [inline themes/cardinal_service/cardinal_service.info.yml:39]: isn't `//` an anti-pattern now? Shouldn't you use `https://` if it is available on the remote resource?

File: themes/cardinal_service/pdb_components/babel.config.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+      //   [
+      //     "transform-react-remove-prop-types",
+      //     {
+      //       removeImport: true
+      //     }
+      //   ],
+      //   "@babel/plugin-transform-react-inline-elements",
+      //   "@babel/plugin-transform-react-constant-elements"
+      // ]
+    }
```
Comment (sherakama) [inline themes/cardinal_service/pdb_components/babel.config.js:29]: Remove commented out code if you are not using it.

File: themes/cardinal_service/cardinal_service.info.yml (hunk 2 above)
Comment (pookmish) [reply themes/cardinal_service/cardinal_service.info.yml:39]: yeah i'm pretty sure you're right

**Comment by sherakama** [comment]
Ok, I validated that the functionality works on my local. One last pass at the code and this should be GTG

File: modules/cardinal_service_rest/cardinal_service_rest.info.yml [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
+version: 8.x-1.x-dev
```
Comment (sherakama) [inline modules/cardinal_service_rest/cardinal_service_rest.info.yml:5]:
```suggestion
version: 8.x-1.0-dev
```

File: themes/cardinal_service/pdb_components/opportunities_list/src/Components/SelectList.jsx [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+export class SelectList extends Component {
+
+  constructor(props) {
+    super(props);
+    this.uuid = _.uuid();
+    this.state = {selectedItems: []};
+    this.onChange = this.onChange.bind(this);
+  }
+
+  onChange(selectedOptions) {
```
Comment (sherakama) [inline themes/cardinal_service/pdb_components/opportunities_list/src/Components/SelectList.jsx:15]: We should probably discuss and define our standards for JSX templates. I feel that each function probably should have a comment.

File: themes/cardinal_service/pdb_components/opportunities_list/src/Components/SelectList.jsx [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  filterOptions = (candidate, input) => {
+    if (input) {
+      return candidate.data.label.toLowerCase().indexOf(input.toLowerCase()) >= 0;
+    }
+    return true;
+  };
+
+  render() {
+    if (this.props.options === undefined) {
+      return <React.Fragment/>
```
Comment (sherakama) [inline themes/cardinal_service/pdb_components/opportunities_list/src/Components/SelectList.jsx:34]: TIL: https://reactjs.org/docs/fragments.html

File: themes/cardinal_service/pdb_components/opportunities_list/src/Components/SelectList.jsx [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    const options = this.props.options.map(option => ({
+      value: option.id,
+      label: `${option.label} (${option.items.length})`,
+      resultCount: option.items.length
+    }))
```
Comment (sherakama) [inline themes/cardinal_service/pdb_components/opportunities_list/src/Components/SelectList.jsx:41]:
```suggestion
    }));
```

File: themes/cardinal_service/pdb_components/opportunities_list/src/Components/SelectList.jsx [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+
+    let value = [];
+    if (this.props.defaultValue !== undefined) {
+      this.props.defaultValue.map(tid => {
+        const option = options.find(opt => parseInt(opt.value) === parseInt(tid));
+        value.push(option)
+      })
+    }
+
+    const className = this.props.field.split('_').join('-')
```
Comment (sherakama) [inline themes/cardinal_service/pdb_components/opportunities_list/src/Components/SelectList.jsx:51]:
```suggestion
    const className = this.props.field.split('_').join('-');
```

File: themes/cardinal_service/pdb_components/opportunities_list/src/Components/SelectList.jsx (hunk 6 above)
Comment (pookmish) [reply themes/cardinal_service/pdb_components/opportunities_list/src/Components/SelectList.jsx:34]: yeah its a very useful thing to keep the markup lower. always gotta have wrapper, but not necessarily a visible one.

File: themes/cardinal_service/pdb_components/opportunities_list/src/Components/SelectList.jsx (hunk 8 above)
Comment (pookmish) [reply themes/cardinal_service/pdb_components/opportunities_list/src/Components/SelectList.jsx]: i should see if codeclimate can catch those

File: themes/cardinal_service/pdb_components/opportunities_list/src/app.jsx [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+
+      if (validEntities.length > 0) {
+        newState.activeItems[fieldName].map(term => {
+          term.items = validEntities.filter(x => term.items.includes(x));
+        })
+      }
+    })
+
+    this.setState(newState);
+  }
```
Comment (sherakama) [inline themes/cardinal_service/pdb_components/opportunities_list/src/app.jsx:106]: This block needs comments to what is happening here. This is complicated enough that it needs explanation.

File: themes/cardinal_service/pdb_components/package.json [hunk 10]
```diff
@@ -0,0 +1,35 @@
+{
+  "name": "pdb_components",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "dev": "webpack --mode development",
+    "build": "webpack --mode production"
```
Comment (sherakama) [inline themes/cardinal_service/pdb_components/package.json:8]: I think we have been using the conventions of `npm/yarn build` for development builds and `npm/yarn publish` for production builds.

File: themes/cardinal_service/pdb_components/package.json [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
```
Comment (sherakama) [inline themes/cardinal_service/pdb_components/package.json:12]: What is ISC?

File: themes/cardinal_service/pdb_components/webpack.config.js [hunk 12]
```diff
@@ -0,0 +1,35 @@
+const path = require("path");
+
+
```
Comment (sherakama) [inline themes/cardinal_service/pdb_components/webpack.config.js:3]: Extra space

File: themes/cardinal_service/pdb_components/opportunities_list/src/Components/SelectList.jsx (hunk 6 above)
Comment (sherakama) [reply themes/cardinal_service/pdb_components/opportunities_list/src/Components/SelectList.jsx:34]: I've used `<>` in the past but I like this better.

**Comment by sherakama** [comment]
Overall this looks fine to me. I would like you to take a pass at commenting the JSX templates with more detail. There is a lot going on in there that should be clarified.

File: themes/cardinal_service/pdb_components/opportunities_list/src/app.jsx (hunk 9 above)
Comment (pookmish) [reply themes/cardinal_service/pdb_components/opportunities_list/src/app.jsx:106]: oh please. its clear cut isn't it. :laughing:

**Comment by sherakama** [comment]
Accessibility and browser testing will come after theming?

**Comment by pookmish** [comment]
yeah. theres some changes that are planned to this as to what filters and what order so theres more to be done as that goes. I did some initial accessibility testing and it has really good results

File: themes/cardinal_service/pdb_components/package.json (hunk 11 above)
Comment (pookmish) [reply themes/cardinal_service/pdb_components/package.json:12]: :shrug: Thats just what happens when i did `npm init -y`

**Comment by pookmish** [comment]
@sherakama I added bunches of comments. can you have a quick once over

**Comment by pookmish** [comment]
one last chance. gonna merge this since its past due and now a bigger blocker

**Pull request merged**